### DISCUSSION TITLE
kex_exchange_identification: Connection closed by remote host

### DISCUSSION THREAD

**Discussion by dongdongbh** [community/community#55269] [New to GitHub]
My git(hub) works fine before, but suddenly it not work.
When I type `git pull origin master`, it shows 
```
kex_exchange_identification: Connection closed by remote host
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```
I have configed key, and it works before. My network works fine. `ping github.com` also works.
Here is the debug message
```
ssh -vT git@github.com
OpenSSH_8.2p1 Ubuntu-4ubuntu0.5, OpenSSL 1.1.1f  31 Mar 2020
debug1: Reading configuration data /home/dd/.ssh/config
debug1: Reading configuration data /etc/ssh/ssh_config
debug1: /etc/ssh/ssh_config line 19: include /etc/ssh/ssh_config.d/*.conf matched no files
debug1: /etc/ssh/ssh_config line 21: Applying options for *
debug1: Connecting to github.com [20.205.243.166] port 22.
debug1: Connection established.
debug1: identity file /home/dd/.ssh/id_rsa type 0
debug1: identity file /home/dd/.ssh/id_rsa-cert type -1
debug1: identity file /home/dd/.ssh/id_dsa type -1
debug1: identity file /home/dd/.ssh/id_dsa-cert type -1
debug1: identity file /home/dd/.ssh/id_ecdsa type -1
debug1: identity file /home/dd/.ssh/id_ecdsa-cert type -1
debug1: identity file /home/dd/.ssh/id_ecdsa_sk type -1
debug1: identity file /home/dd/.ssh/id_ecdsa_sk-cert type -1
debug1: identity file /home/dd/.ssh/id_ed25519 type -1
debug1: identity file /home/dd/.ssh/id_ed25519-cert type -1
debug1: identity file /home/dd/.ssh/id_ed25519_sk type -1
debug1: identity file /home/dd/.ssh/id_ed25519_sk-cert type -1
debug1: identity file /home/dd/.ssh/id_xmss type -1
debug1: identity file /home/dd/.ssh/id_xmss-cert type -1
debug1: Local version string SSH-2.0-OpenSSH_8.2p1 Ubuntu-4ubuntu0.5
kex_exchange_identification: Connection closed by remote host
```

```
git remote -v
origin  git@github.com:dongdongbh/dotfiles.git (fetch)
origin  git@github.com:dongdongbh/dotfiles.git (push)
```

**Comment by airtower-luna** [comment]
The SSH output looks like you don't have any keys available for authentication. Do you have SSH-Agent running, and the right key pair loaded into it?

**Comment by xc2f** [comment]
I encountered the same problem 3 hours ago, unable to push the code to github, it showed no permission, but it was still able to push normally just yesterday. After I generate a new ssh key locally and add it to the github ssh configuration, this problem still exists.
After about five or six minutes, I tried to push again, and the push was successful.

A strange thing happened to me the other day when I was syncing code on another computer. The remote branch is one commit ahead of the local branch, but when I execute `git pull` or `git fetch / rebase` locally, it shows `Already up to date.`. Finally, I deleted the local project and re-pulled the remote code to solve the problem.

**Reply by LeisureLinux** [reply]
I still have the same issue here in Shanghai. git pull/push through proxy though. 
But apparently, since it is through ssh protocol, then the http.proxy setup should not affect.

**Comment by dongdongbh** [comment]
I have sshd daemon running. I didn't make any change, and now it works, so strange~

**Reply by LeisureLinux** [reply]
Please leave this open. It is still not working for me here in Shanghai.

**Reply by LeisureLinux** [reply]
I guess I found the root cause, and had a work around by adding a line in my /etc/hosts Linux Desktop. 
pick an IP address from git section in https://api.github.com/meta 

❯ grep github /etc/hosts                 
20.201.28.152 github.com


my original resolving github.com has address 20.205.243.166, is not working. 

Unfortunately, no service outage reported in githubstatus.com page.

**Reply by LeisureLinux** [reply]
Just after a working git pull, and re-run git pull again, found same issue happened on the new IP. 
So this workaround won't work, since it is a server side issue.

**Reply by xuthus5** [reply]
> I guess I found the root cause, and had a work around by adding a line in my /etc/hosts Linux Desktop. pick an IP address from git section in https://api.github.com/meta
> 
> ❯ grep github /etc/hosts 20.201.28.152 github.com
> 
> my original resolving github.com has address 20.205.243.166, is not working.
> 
> Unfortunately, no service outage reported in githubstatus.com page.

thank you, good for me.

**Reply by 1990jtf** [reply]
wow~ It works!  Thanks a lot!

**Reply by cooljacket** [reply]
But it cause that when i access github.com in Chrome, the web page cannot display right. Finally i recover the local host dns config, use the following LeisureLinux's solution.

**Reply by euskate** [reply]
thanks~ work for me

**Reply by BobJackson** [reply]
I just pick a ip from [https://api.github.com/meta]() , write to host file, it works!

**Comment by LeisureLinux** [comment]
I guess GFW make it. 
By checking: https://docs.github.com/en/authentication/troubleshooting-ssh/using-ssh-over-the-https-port 
which I followed the ~/.ssh/config added the port 443.
and use command like: "proxychains -q git pull", it is working now.

Here is what is working now through https, and through proxy: 

❯ proxychains -q  ssh -T git@github.com
The authenticity of host '[ssh.github.com]:443 ([20.205.243.160]:443)' can't be established.
ED25519 key fingerprint is SHA256:+DiY3wvvV6TuJJhbpZisF/zLDA0zPMSvHdkr4UvCOqU.
This key is not known by any other names
Are you sure you want to continue connecting (yes/no/[fingerprint])? yes
Warning: Permanently added '[ssh.github.com]:443' (ED25519) to the list of known hosts.
Hi LeisureLinux! You've successfully authenticated, but GitHub does not provide shell access.

**Reply by dongdongbh** [reply]
Hi, I'm also in Shanghai. Thanks for your feedback. I also guess it related to GFW. I tested it yesterday when it not working. If I got lucky in a trial, it worked. I also tried write 443 to .ssh/config, it stuck in my case.

**Reply by flaymes** [reply]
I have the same problem, I'm in Guangzhou

**Reply by flaymes** [reply]
> I have the same problem, I'm in Guangzhou

I just deleted  github.com record in known_hosts  file, It is work for me.

**Reply by dongdongbh** [reply]
Seems it may has problem in some areas of China. I guess it related to the network condition of China.

**Reply by weiwongfaye** [reply]
same here, shanghai, China Telecommunications, `kex_exchange_identification: Connection closed by remote host`
switch to phone hot pot, then it is fine. 

also, connecting to aws ec2 also very slow in recent days

**Reply by renwuli** [reply]
Thank you! You saved my life!

**Reply by JTR354** [reply]
Thank you! it's working, have a nice time.

**Reply by AtaraxiaZ** [reply]
Thank you! Now I could use the `export http_proxy=http://127.0.0.1:7890` in terminal from Clash proxy and let git operate on https.

**Reply by Nu11able** [reply]
It works! Thank you so much

**Reply by lxd6701996** [reply]
macOS 13.6.3 is ok

**Reply by ZiyangQian** [reply]
thanks so much

**Reply by charsm** [reply]
it's working ~~~. nice

**Reply by stupidchen** [reply]
Cool! It works for me. Thanks!

**Reply by HoganBright** [reply]
Thank you so much!
2024.05.13 Mon 12:45 from China

**Reply by VanillaChelle** [reply]
It doesn't work for me by changing ssh to https. I met the same `kex_exchange_identification: Connection closed by remote host` issue with github ip 20.205.243.166, both in Windows and Mac. Changing ip address resolves this issue. Really appreciate it!

**Reply by wangy325** [reply]
Thx a lot! works for me.

**Reply by ttian20** [reply]
It works for me. Thanks very much.  Changing ip address in /etc/hosts not works.

`Host github.com
    Hostname ssh.github.com
    Port 443
    User git`

**Comment by dongdongbh** [comment]
Because many people in China encounter this problem, seems it's related to network. I reopen this issue. More solution and feedback are welcome.

**Comment by LeslieLeung** [comment]
Try `ssh-keyscan github.com >> ~/.ssh/known_hosts`. Works for me.

Found on this [post](https://www.reddit.com/r/github/comments/dxej4m/ssh_exchange_identification_connection_closed_by/)

**Reply by airtower-luna** [reply]
Never do this, *especially* not if you know or suspect the government is behind the problem. :warning:

It means you accept whatever keys the server offers, allowing them to intercept your connection. If there's a problem with the public keys of the server, check if the fingerprint you're seeing [is one of the published ones](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/githubs-ssh-key-fingerprints). Same for any other SSH server.

**Reply by LeslieLeung** [reply]
Thanks for the notice. If anyone did the above, just use `ssh-keygen -R github.com` to undo it.

P.S. I've checked my known_hosts, the fingerprints are identical to the published ones. Maybe whoever's doing this is not clever/malicious enough to do this.

**Reply by AffordCK** [reply]
It works for me, thanks!

**Comment by jinfanchris** [comment]
Same here in Tianjin, and it's resolved by executing the following command, found [here](https://docs.github.com/en/authentication/troubleshooting-ssh/using-ssh-over-the-https-port).
```
ssh -T -p 443 git@ssh.github.com
```
It seems that something, related to ssh identification or Github ip resolution, has been changed during the execution.

**Reply by Hjh233** [reply]
Thanks a lot！Works for me.

**Reply by neufeng** [reply]
Very good! Works for me too! I'm in ShangHai! Thanks a lot!

**Reply by berrytsukiii** [reply]
thx！

**Reply by Angeladadd** [reply]
Thanks! Works for me along with closing VPN.

**Reply by zhaziqwe** [reply]
thank you， it really works

**Reply by swahpy** [reply]
this works for me. Together with adding IP to `/etc/hosts` file by checking https://dnschecker.org/#A/ssh.github.com. Thanks a lot!

**Comment by Seraphli** [comment]
Adding this to `~/.ssh/config` solves my problem
```
Host github.com
  Hostname 20.200.245.248
  Port 443
```
I use South Korea IP.
https://dnschecker.org/#A/ssh.github.com

**Reply by Seraphli** [reply]
And then I delete this part and it still works. I guess GFW is causing this problem.

**Reply by bitcapybara** [reply]
thx, this works for me!!

**Reply by AlphaChen** [reply]
thx it works for me

**Reply by Miiiiiiir** [reply]
works for me,thx a lot

**Reply by Sugayoiya** [reply]
works

**Reply by cheny-00** [reply]
works for me! thx for sharing

**Reply by 08822407d** [reply]
works! you are the hero!

**Reply by MasonSyj** [reply]
Works for me as well. I'm at Shanghai

**Reply by Hajduk137** [reply]
works for me, thanks a lot

**Reply by chaseSpace** [reply]
This method clearly illustrates that github.com experienced a significant DNS pollution issue in China, which is caused by GFW.

**Reply by seec47x** [reply]
Works for me, really appreciate your help

**Reply by GreyRaphael** [reply]
Works for me, really appreciate your help

**Reply by jackwang0108** [reply]
Cool, this works for me, appreciated ;)

**Reply by SWHL** [reply]
works for me. Thanks a lot! 
I am in Beijing.

**Reply by mileschen0419** [reply]
Save my life!!!

**Reply by yiwenup** [reply]
thx!

**Reply by andykimisgreat** [reply]
It worked. Thanks!

**Reply by kelinkong** [reply]
Thanks!

**Reply by bqdcc** [reply]
Thanks!

**Reply by chenyang-zheng** [reply]
Life saver, really appreciate it!

**Reply by yemaobumei** [reply]
it works for me. but why???

**Comment by leeeezp** [comment]
I have the same problem. It will be normal when I don't use clash.

**Comment by iDestin** [comment]
I am in Beijing and recently discovered that SSH is not working properly, but I have not found a solution yet

**Comment by HxYxI** [comment]
> My git(hub) works fine before, but suddenly it not work.
> 
> When I type `git pull origin master`, it shows 
> 
> ```
> 
> kex_exchange_identification: Connection closed by remote host
> 
> fatal: Could not read from remote repository.
> 
> 
> 
> Please make sure you have the correct access rights
> 
> and the repository exists.
> 
> ```
> 
> I have configed key, and it works before. My network works fine. `ping github.com` also works.
> 
> Here is the debug message
> 
> ```
> 
> ssh -vT git@github.com
> 
> OpenSSH_8.2p1 Ubuntu-4ubuntu0.5, OpenSSL 1.1.1f  31 Mar 2020
> 
> debug1: Reading configuration data /home/dd/.ssh/config
> 
> debug1: Reading configuration data /etc/ssh/ssh_config
> 
> debug1: /etc/ssh/ssh_config line 19: include /etc/ssh/ssh_config.d/*.conf matched no files
> 
> debug1: /etc/ssh/ssh_config line 21: Applying options for *
> 
> debug1: Connecting to github.com [20.205.243.166] port 22.
> 
> debug1: Connection established.
> 
> debug1: identity file /home/dd/.ssh/id_rsa type 0
> 
> debug1: identity file /home/dd/.ssh/id_rsa-cert type -1
> 
> debug1: identity file /home/dd/.ssh/id_dsa type -1
> 
> debug1: identity file /home/dd/.ssh/id_dsa-cert type -1
> 
> debug1: identity file /home/dd/.ssh/id_ecdsa type -1
> 
> debug1: identity file /home/dd/.ssh/id_ecdsa-cert type -1
> 
> debug1: identity file /home/dd/.ssh/id_ecdsa_sk type -1
> 
> debug1: identity file /home/dd/.ssh/id_ecdsa_sk-cert type -1
> 
> debug1: identity file /home/dd/.ssh/id_ed25519 type -1
> 
> debug1: identity file /home/dd/.ssh/id_ed25519-cert type -1
> 
> debug1: identity file /home/dd/.ssh/id_ed25519_sk type -1
> 
> debug1: identity file /home/dd/.ssh/id_ed25519_sk-cert type -1
> 
> debug1: identity file /home/dd/.ssh/id_xmss type -1
> 
> debug1: identity file /home/dd/.ssh/id_xmss-cert type -1
> 
> debug1: Local version string SSH-2.0-OpenSSH_8.2p1 Ubuntu-4ubuntu0.5
> 
> kex_exchange_identification: Connection closed by remote host
> 
> ```
> 
> 
> 
> ```
> 
> git remote -v
> 
> origin  git@github.com:dongdongbh/dotfiles.git (fetch)
> 
> origin  git@github.com:dongdongbh/dotfiles.git (push)
> 
> ```
</*> .

**Comment by spianmo** [comment]
Enabling SSH connections over HTTPS can solve this problem
If you are able to SSH into [git@ssh.github.com](mailto:git@ssh.github.com) over port 443, you can override your SSH settings to force any connection to GitHub.com to run through that server and port.

To set this in your SSH configuration file, edit the file at ~/.ssh/config, and add this section:

```
Host github.com
    Hostname ssh.github.com
    Port 443
    User git
```

You can test that this works by connecting once more to GitHub.com:

$ ssh -T [git@github.com](mailto:git@github.com)

> Hi USERNAME! You've successfully authenticated, but GitHub does not
> provide shell access.

**Reply by OffixialK** [reply]
Host github.com
    Hostname ssh.github.com
    Port 443
    User git

**Reply by stevenlii** [reply]
git@ssh.github.com: Permission denied (publickey).

**Reply by liang7hi** [reply]
it works for me

**Reply by zerone0x** [reply]
It works!!!

**Reply by ellutionist** [reply]
Works, g damn GFW.

**Reply by unicornZhen** [reply]
thanks!!!

**Reply by zerone0x** [reply]
It happened again, and works again!

**Reply by rudy776649** [reply]
For me, it works. To make ssh to use proxy at local, I find out solution like this.

```
ProxyCommand "C:\APP\Git\mingw64\bin\connect" -S 127.0.0.1:51837 -a none %h %p

Host github.com
  User git
  Port 22
  Hostname github.com
  IdentityFile "C:\Users\Your_User_Name\.ssh\id_rsa"
  TCPKeepAlive yes

Host ssh.github.com
  User git
  Port 443
  Hostname ssh.github.com
  IdentityFile "C:\Users\Your_User_Name\.ssh\id_rsa"
  TCPKeepAlive yes
```

However, it does not work and when I changed as https://docs.github.com/en/authentication/troubleshooting-ssh/using-ssh-over-the-https-port metioned, it works. Change the "Host github.com" part.
```
Host github.com
  User git
  Port 443
  Hostname ssh.github.com
  IdentityFile "C:\Users\Your_User_Name\.ssh\id_rsa"
  TCPKeepAlive yes
```

**Reply by cath-zhou** [reply]
Thanks!! Solves my issue :)

**Reply by Kaviilee** [reply]
thanks,it works for me!

**Reply by fuqiang-repository** [reply]
Thanks! a timely help

**Reply by yaolynzc** [reply]
Thanks,it works for me!

**Reply by cyssxt** [reply]
Thanks!it works for me!

**Reply by Mrhs121** [reply]
万分感谢！！！

**Reply by wangxdmm** [reply]
You save my life....

**Reply by xxxxinc** [reply]
好使的！

**Reply by wa008** [reply]
works for me

**Reply by VictorChow** [reply]
GREAT JOB!!!

**Reply by fedorlong** [reply]
老哥真靠谱，上chatgpt查了下，应该是vpn导致的，需要加443端口

**Reply by Aniu456** [reply]
牛逼！

**Comment by zachx121** [comment]
try 'ssh-agent  && ssh-add ~/.ssh/id_rsa'

**Reply by ning2510** [reply]
This works for me, thx

**Comment by Falasool** [comment]
If you use Proxy client like ClashX, please close the "Enhanced Mode" and try again.

**Reply by Div1to** [reply]
This works for me, thx

**Reply by crazy-dadong** [reply]
This works for me, thx

**Reply by cheung-chifung** [reply]
This works for me, thx

**Reply by iFTY-R** [reply]
This works for me, thx

**Comment by Elon-Lee** [comment]
Becuase of this problem , check your proxy config . Just like  clashx shadowsocks etc ...

**Comment by ccsuper1024** [comment]
sometimes,Maybe ,we should check our .ssh/  , the public key's and private key's file name.it must  start with id_rsa. do not start with anyelse word. I modify the filename ,then ,everything go correct

**Comment by mark-wang-beep** [comment]
I just closed my ClashX, and then problem solved

**Comment by changbindu** [comment]
Check the solution here.
https://stackoverflow.com/questions/10127818/ssh-exchange-identification-connection-closed-by-remote-host-under-git-bash

**Reply by philipjoep** [reply]
This is literally the only thing that worked for me. I'm running a Linux VM on Windows. I unplugged my LAN cable, switched to Wifi, and suddenly I was able to connect to github again. Then after switching back to LAN, it all worked.

Now, I did also update my .ssh/config file and cleared out my known_hosts file before these steps, but they didn't have any affect. I did leave the config file the same.

```
Host github.com
    Hostname ssh.github.com
    Port 443
    User git
```

**Comment by treblam** [comment]
I just closed openclash in my openwrt router, and it works.

**Reply by Marerc** [reply]
Then how can you access github.com?

**Reply by treblam** [reply]
github is not blocked in China, but it’s very slow without using proxy.

**Comment by arko-svmx** [comment]
Make sure if you are using any client connector app like Zscaler, turn its internet security off. It solved every problem for me.
<img width="588" alt="Screenshot 2024-02-21 at 12 49 23 PM" src="https://github.com/community/community/assets/159111330/4cddd9bb-2fc0-4192-bddc-5e46a606faa0">

**Comment by queenofcorgis** [comment]
Given the popularity of this discussion and the quality of information shared, we want to make sure others can easily find it so we’ve reopened this discussion for better discoverability for our members.

**Comment by thesilentpsycho** [comment]
my org vpn was off. please check

**Comment by githika99** [comment]
I added this to ~/.ssh/config and it finally worked! I think the hostname is what makes it work. 
Host github.com
    HostName ssh.github.com
    User git
    Port 443

Source: https://stackoverflow.com/questions/10127818/ssh-exchange-identification-connection-closed-by-remote-host-under-git-bash

**Reply by gospodnetic** [reply]
Adding `Port 443` and changing `HostName` to `HostName ssh.github.com` worked for me! I have user set to my username.

**Comment by Narrator** [comment]
Hi :wave: 

I'm an engineer from the AWS Amplify service - Our customers use GitHub to connect their applications to Amplify and we offer a CI/CD workflow where we clone the customer's repository to allow customers to run their build commands.

Since 09/11, our customers have been intermittently observing this issue:
```
 kex_exchange_identification: Connection closed by remote host
 Connection closed by 20.205.243.166 port 22
 fatal: Could not read from remote repository.
 Please make sure you have the correct access rights
 and the repository exists.
```

The fact that these failures are intermittent is what is concerning to us. Nothing has changed in the build containers we provision for our customers - The SSH settings are the same as they always were.

We run `ssh-keyscan github.com` and add the host keys to the `~/.ssh/known_hosts` file every time before we clone. Our logs show that this call succeeded, but the failure still occurs.

**Reply by wangy325** [reply]
This article may help: https://docs.github.com/en/authentication/troubleshooting-ssh/using-ssh-over-the-https-port#enabling-ssh-connections-over-https

I encountered  same problem on my git suddenly, and fixed it by following the guide.

**Reply by devinburnette** [reply]
@Narrator we also started experiencing this issue intermittently around that same time. Let me know if you find a solution!

**Reply by Narrator** [reply]
I created a separate discussion in the [General](https://github.com/orgs/community/discussions/categories/general) category with  `Bug` label  since it looks very much like something changed on GitHub's side on 09/11: https://github.com/orgs/community/discussions/138757

_(This discussion is in the [New to GitHub](https://github.com/orgs/community/discussions/categories/new-to-github) category and perhaps that's the reason it's not gaining much traction from the GitHub staff)._
